ee/clickhouse/queries/clickhouse_retention.py

```diff
@@ -79,3 +81,75 @@ def _get_trunc_func_ch(self, period: str) -> str:
+            .where(person_distinct_id.team_id == PyformatParameter("team_id"))
+        )
+        return q
+
+    def reference_query(self, trunc: str):
+        events = Table("events").as_("e")
+        person_query = self.person_query().as_("pdi")
+        toDateTime = CustomFunction("toDateTime", ["date"])
+        truncFunc = CustomFunction(trunc, ["date"])
+        return (
+            Query.from_(events)
+            .join(person_query)
+            .on(person_query.distinct_id == events.distinct_id)
+            .select(truncFunc(events.timestamp).as_("event_date"), person_query.person_id.as_("person_id"),)
+            .distinct()
+            .where(toDateTime(events.timestamp) >= PyformatParameter("start_date"))
+            .where(toDateTime(events.timestamp) <= PyformatParameter("end_date"))
+            .where(events.team_id >= PyformatParameter("team_id"))
+        )
+
+    def event_query(self):
+        events = Table("events").as_("e")
+        person_query = self.person_query().as_("pdi")
+        toDateTime = CustomFunction("toDateTime", ["date"])
+
+        return (
+            Query.from_(events)
+            .join(person_query)
+            .on(person_query.distinct_id == events.distinct_id)
+            .select(events.timestamp.as_("event_date"), person_query.person_id.as_("person_id"),)
+            .where(toDateTime(events.timestamp) >= PyformatParameter("start_date"))
+            .where(toDateTime(events.timestamp) <= PyformatParameter("end_date"))
+            .where(events.team_id >= PyformatParameter("team_id"))
+        )
+
+    def final_query(self, trunc: str):
+        toDateTime = CustomFunction("toDateTime", ["date"])
+        truncFunc = CustomFunction(trunc, ["date"])
+        dateDiff = CustomFunction("datediff", ["period", "start_date", "end_date"])
+
+        event_query = self.event_query().as_("event")
+        reference_event = self.reference_query(trunc).as_("reference_event")
+
+        final_query = (
+            Query.from_(event_query)
+            .join(reference_event)
+            .on(event_query.person_id == reference_event.person_id)
+            .select(
+                dateDiff(
+                    PyformatParameter("period"),
+                    truncFunc(toDateTime(PyformatParameter("start_date"))),
+                    reference_event.event_date,
+                ).as_("period_to_event_days"),
+                dateDiff(
+                    PyformatParameter("period"),
+                    reference_event.event_date,
+                    truncFunc(toDateTime(event_query.event_date)),
+                ).as_("period_between_events_days"),
+                fn.Count(event_query.person_id).distinct().as_("count"),
+            )
+            .where(truncFunc(event_query.event_date) >= truncFunc(reference_event.event_date))
+            .groupby(Field("period_to_event_days"), Field("period_between_events_days"))
+        )
```

Comment:
This is what the retention query would look like. It allows custom function definitions so this would allow swapping in and out postgres/clickhouse table names/special functions